src/bun.js/api/bun/udp_socket.zig

```diff
@@ -0,0 +1,714 @@
+const std = @import("std");
+const uws = @import("../../../deps/uws.zig");
+const bun = @import("root").bun;
+
+const strings = bun.strings;
+const default_allocator = bun.default_allocator;
+const Output = bun.Output;
+const Async = bun.Async;
+const JSC = bun.JSC;
+const CallFrame = JSC.CallFrame;
+const JSGlobalObject = JSC.JSGlobalObject;
+const JSValue = JSC.JSValue;
+
+const log = Output.scoped(.UdpSocket, false);
+
+const INET6_ADDRSTRLEN = if (bun.Environment.isWindows) 65 else 46;
+
+extern fn ntohs(nshort: u16) u16;
+extern fn htonl(hlong: u32) u32;
+extern fn htons(hshort: u16) u16;
+extern fn inet_ntop(af: c_int, src: ?*const anyopaque, dst: [*c]u8, size: c_int) ?[*:0]const u8;
+extern fn inet_pton(af: c_int, src: [*c]const u8, dst: ?*anyopaque) c_int;
+extern fn JSSocketAddress__create(global: *JSGlobalObject, address: JSValue, port: i32, v6: bool) JSValue;
+
+fn onDrain(socket: *uws.udp.Socket) callconv(.C) void {
+    JSC.markBinding(@src());
+
+    const this: *UDPSocket = bun.cast(*UDPSocket, socket.user().?);
+    const callback = this.config.on_drain;
+    if (callback == .zero) return;
+
+    const result = callback.callWithThis(this.globalThis, this.thisValue, &[_]JSValue{this.thisValue});
+    if (result.toError()) |err| {
+        _ = this.callErrorHandler(.zero, &[_]JSValue{err});
+    }
+}
+
+fn onData(socket: *uws.udp.Socket, buf: *uws.udp.PacketBuffer, packets: c_int) callconv(.C) void {
+    JSC.markBinding(@src());
+
+    const udpSocket: *UDPSocket = bun.cast(*UDPSocket, socket.user().?);
+    const callback = udpSocket.config.on_data;
+    if (callback == .zero) return;
+
+    const globalThis = udpSocket.globalThis;
+
+    var i: c_int = 0;
+    while (i < packets) : (i += 1) {
+        const peer = buf.getPeer(i);
+
+        var addr_buf: [INET6_ADDRSTRLEN + 1:0]u8 = undefined;
+        var hostname: ?[*:0]const u8 = null;
+        var port: u16 = 0;
+
+        switch (peer.family) {
+            std.os.AF.INET => {
+                const peer4: *std.os.sockaddr.in = @ptrCast(peer);
+                hostname = inet_ntop(peer.family, &peer4.addr, &addr_buf, addr_buf.len);
+                port = ntohs(peer4.port);
+            },
+            std.os.AF.INET6 => {
+                const peer6: *std.os.sockaddr.in6 = @ptrCast(peer);
+                hostname = inet_ntop(peer.family, &peer6.addr, &addr_buf, addr_buf.len);
+                port = ntohs(peer6.port);
+            },
+            else => continue,
+        }
+
+        if (hostname == null or port == 0) {
+            continue;
+        }
+
+        const slice = buf.getPayload(i);
+
+        const loop = udpSocket.vm.eventLoop();
+        loop.enter();
+        defer loop.exit();
+        udpSocket.js_refcount += 1;
```

Comment:
this feels strange imo

also if we are using has_pending_activity, then this needs to be done threadsafely which means using a `std.atomic.Value` and calling fetchadd/fetchsub